### PR TITLE
[WIP] Use pip-accel for building jenkins workers

### PR DESCRIPTION
**Pull request by jzoldak**
@benpatterson @fredsmith FYI

### CODE REVIEW

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 1]
```diff
@@ -30,18 +30,50 @@
...
+# rather than compiled each time.
+- name: Install pip-accel for caching python requirements
+  pip: >
+    name='git+https://github.com/paylogic/pip-accel.git@0.18#egg=pip-accel'
+    virtualenv={{ jenkins_home }}/edx-venv
+    virtualenv_command=virtualenv-2.7
+  sudo_user: "{{ jenkins_user }}"
+
+# pip-accel only makes the s3 functionality available
+# if you have installed boto yourself.
```
Comment (benpatterson) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: Is this because you need the .boto file? Perhaps we just add the .boto file and install boto via the usual requirements? Would that work?

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 2]
```diff
@@ -30,18 +30,50 @@
...
+- name: Install boto for caching python requirements
+  pip: >
+    name='boto' version='2.34.0'
```
Comment (benpatterson) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: FWIW this version number is different than platform right now. I think when we run the installs on platform we would need to use the wipe setting if we don't already.

File: util/packer/README.rst [hunk 3]
```diff
@@ -0,0 +1,29 @@
+Packer
```
Comment (benpatterson) [inline util/packer/README.rst:1]: Thanks a lot for adding doc.

File: util/packer/jenkins_worker.json [hunk 4]
```diff
@@ -35,7 +41,7 @@
       "pip install -q -r requirements.txt",
       "echo '[jenkins_worker]' > inventory.ini",
       "echo 'localhost' >> inventory.ini",
-      "ansible-playbook edx-east/jenkins_worker.yml -i inventory.ini -c local -e 'github_oauth_token={{user `github_oauth_token` }}' -vvvv"]
+      "ansible-playbook edx-east/jenkins_worker.yml -i inventory.ini -c local -e '@extra_vars.json' -vvvv"]
```
Comment (benpatterson) [inline util/packer/jenkins_worker.json]: I like this particular change as well.

File: playbooks/roles/jenkins_worker/tasks/python.yml (hunk 1 above)
Comment (jzoldak) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: No, you need more than just the .boto file. You need to have boto installed in the virtualenv for pip-accel to take advantage of the s3 cache, and the maintainer of pip-accel didn't want to include boto as a requirement. So you either need to install it yourself, or you need to install pip-accel like this: `pip install pip-accel[s3]`

File: playbooks/roles/jenkins_worker/tasks/python.yml (hunk 2 above)
Comment (jzoldak) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: ah, good one. I will need to figure out what to pass through to pip-accel as an extra arg then, I had briefly tried it but it was giving some error message that I didn't feel like tracking down at that time. :(

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 5]
```diff
@@ -30,18 +30,50 @@
     depth=1
   sudo_user: "{{ jenkins_user }}"
 
-- name: Install edx-platform requirements
+# Install the platform requirements using pip-accel.
+# This will allow the binary distributions to be downloaded from S3
+# rather than compiled each time.
+- name: Install pip-accel for caching python requirements
+  pip: >
+    name='git+https://github.com/paylogic/pip-accel.git@0.18#egg=pip-accel'
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: Can we install this from PyPI instead of github please.  https://pypi.python.org/pypi/pip-accel/0.19

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 6]
```diff
@@ -30,18 +30,50 @@
...
+  pip: >
+    name='boto' version='2.34.0'
+    virtualenv={{ jenkins_home }}/edx-venv
+    virtualenv_command=virtualenv-2.7
+  sudo_user: "{{ jenkins_user }}"
+
+- name: Install edx-platform requirements using pip-accel
   pip: >
     requirements={{ jenkins_home }}/shallow-clone/requirements/edx/{{ item }}
-    extra_args="--exists-action w"
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/python.yml:36]: Why was this removed?

**Comment by feanil** [comment]
Why does pip need AWS credentials?

File: playbooks/roles/jenkins_worker/tasks/python.yml (hunk 2 above)
Comment (jzoldak) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: Update: figured out the wipe exists-action. This boto install should be fine now.

File: playbooks/roles/jenkins_worker/tasks/packer.yml [hunk 7]
```diff
@@ -1,6 +1,8 @@
 ---
 - name: Download packer
   shell: "curl -L {{ packer_url }} -o /var/tmp/packer.zip"
+  args:
```
Comment (jzoldak) [inline playbooks/roles/jenkins_worker/tasks/packer.yml:4]: review note: this is unrelated to the other changes. It just speeds up the "vagrant provision" when I was developing.

File: playbooks/roles/jenkins_worker/tasks/python.yml (hunk 6 above)
Comment (jzoldak) [inline playbooks/roles/jenkins_worker/tasks/python.yml:36]: Was stuck on that. Turned out to be a bug in pip 1.4.1. Added back in. 

Also I'm not sure why there was no `=` sign originally, but I have confirmed that it works with it.

**Comment by jzoldak** [comment]
@benpatterson @feanil 
All comments addressed. Please re-review.
Looking to merge this in.

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 8]
```diff
@@ -30,12 +30,61 @@
     depth=1
   sudo_user: "{{ jenkins_user }}"
 
-- name: Install edx-platform requirements
+# Install the platform requirements using pip-accel.
+# This will allow the binary distributions to be downloaded from S3
+# rather than compiled each time.
+- name: Install pip-accel for caching python requirements
+  pip: >
+    name='pip-accel' version='0.21.1'
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: I would rather see a var that defines all the reqeuirements that need to be installed in defaults along with their versions, and then a single task to install both boto and pip-accel.  This will keep things DRY and make it easier to add other requirements in the future.

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 9]
```diff
@@ -30,12 +30,61 @@
...
+    name='boto' version='2.34.0'
+    virtualenv={{ jenkins_home }}/edx-venv
+    virtualenv_command=virtualenv-2.7
+  sudo_user: "{{ jenkins_user }}"
+
+# WORKAROUND for pip-accel
+# Install a version of pip that will work with the exists-action flag
+- name: Install pip 1.4.1 patched
+  pip: >
+    name='git+https://github.com/jzoldak/pip.git@v1.4.1patch772#egg=pip'
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: Is there an issue for this berakage that we can also link to here?  What is the plan to not have to do this in the future?  Is this a bug in pip 1.5.6?  More comments here would be generally useful.

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 10]
```diff
@@ -30,12 +30,61 @@
...
+    virtualenv={{ jenkins_home }}/edx-venv
+    virtualenv_command=virtualenv-2.7
+  sudo_user: "{{ jenkins_user }}"
+
+# WORKAROUND for pip-accel
+# Install Shapely which doesn't install cleanly with pip-accel
+- name: Install Shapely==1.2.16
+  pip: >
+    name='Shapely' version='1.2.16'
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: this can also go into the list of pre_pip_accel requirements.

File: playbooks/roles/jenkins_worker/tasks/python.yml [hunk 11]
```diff
@@ -30,12 +30,61 @@
...
+    virtualenv={{ jenkins_home }}/edx-venv
+    virtualenv_command=virtualenv-2.7
+    executable=pip
+  sudo_user: "{{ jenkins_user }}"
+
+# WORKAROUND for pip-accel
+# Install unittest2 which is needed by lettuce
+- name: Install unittest2
+  pip: >
+    name='unittest2'
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/python.yml]: same as other pre pip-accel requirements.  Also why doesn't this have a version?

File: playbooks/roles/jenkins_worker/tasks/system.yml [hunk 12]
```diff
@@ -57,3 +63,19 @@
...
+
+- name: Create pip-accel config file
+  template:
+    src=pip-accel.conf.j2 dest={{ jenkins_home }}/.pip-accel/pip-accel.conf
+    owner={{ jenkins_user }} group={{ jenkins_group }} mode=0664
+
+# Set up configuration for pip-accel to use S3
+- name: Create boto config file with AWS credentials for the pip cache
+  template:
+    src=.boto.j2 dest={{ jenkins_home }}/.boto
```
Comment (feanil) [inline playbooks/roles/jenkins_worker/tasks/system.yml]: If pip accel is using boto, why not use IAM instead of passing around credentials?

File: playbooks/roles/jenkins_worker/tasks/system.yml [hunk 13]
```diff
@@ -34,6 +34,12 @@
   apt: pkg={{','.join(jenkins_debian_pkgs)}}
        state=present update_cache=yes
 
+- fail: AWS access key id was not passed in
```
Comment (benpatterson) [inline playbooks/roles/jenkins_worker/tasks/system.yml]: Might want to specify which kind of key (since we have different S3 keys and we have EC2 keys...)

File: playbooks/roles/jenkins_worker/tasks/system.yml [hunk 14]
```diff
@@ -34,6 +34,12 @@
...
+  when: pip_aws_access_key_id is not defined
+
+- fail: AWS secret access key was not passed in
```
Comment (benpatterson) [inline playbooks/roles/jenkins_worker/tasks/system.yml]: Same here

File: util/packer/jenkins_worker.json [hunk 15]
```diff
@@ -28,6 +30,10 @@
     "source": "../../requirements.txt",
     "destination": "{{user `playbook_remote_dir`}}/requirements.txt"
   }, {
+    "type": "file",
+    "source": "extra_vars.json",
```
Comment (benpatterson) [inline util/packer/jenkins_worker.json]: Does this need to be added to the commit? Or, built by packer before this step?

Comment (benpatterson) [inline util/packer/jenkins_worker.json]: Oh I see now...maybe this file provisioner should just be removed?

**Comment by jzoldak** [comment]
@benpatterson passed the jenkins build-jenkins-worker packer tests. All comments addressed. Can I get a final review pls?

File: playbooks/roles/jenkins_worker/templates/pip-accel.conf.j2 [hunk 16]
```diff
@@ -0,0 +1,7 @@
+[pip-accel]
+auto-install = no
+data-directory = ~/.pip-accel
+download-cache = ~/.pip/download-cache
+s3-bucket = edx-platform.pip-accel-cache
```
Comment (benpatterson) [inline playbooks/roles/jenkins_worker/templates/pip-accel.conf.j2:5]: should 'edx-platform.pip-accel-cache' be a var (so non-platform implementations can use)?

**Comment by benpatterson** [comment]
LGTM. :+1: 

I have a nit, but not a must before merging.

File: playbooks/roles/jenkins_worker/templates/pip-accel.conf.j2 (hunk 16 above)
Comment (jzoldak) [inline playbooks/roles/jenkins_worker/templates/pip-accel.conf.j2:5]: good point. will refactor on the next go-around.
